simplemma/language_detector.py

```diff
@@ -87,42 +76,102 @@ def lang_detector(
-        # compute results
-        found_ratio = in_target / total_tokens
-        myresults[lang_code] = found_ratio
-        unknown = 1 - found_ratio or 0.0
-        if myresults.get("unk") is None or unknown < myresults["unk"]:
-            myresults["unk"] = unknown
-    results = sorted(myresults.items(), key=itemgetter(1), reverse=True)
+    """Determine which proportion of the text is in the target language(s)."""
+    results = LanguageDetector(
+        dictionary_factory, token_sampler
+    ).get_text_percentage_in_each_language(text, lang, greedy)
     # post-processing
-    if len(results) > 1:
-        # switch unknown to the end
-        for i, item in enumerate(results):
-            if item[0] == "unk":
-                pair = results.pop(i)
-                results.append(pair)
-        # in case of ex-aequo use other token sampling to discriminate
-        if not greedy and results[0][1] == results[1][1]:
-            results = lang_detector(
-                text, lang=lang, greedy=True, token_sampler=backup_sampler
-            )
-    return results
+    if len(results) == 1:
+        return _convert_results_to_sorted_list(results)
+    # in case of ex-aequo
+    list_results = _convert_results_to_sorted_list(results)
+    if greedy is False and list_results[0][1] == list_results[1][1]:
+        results = LanguageDetector(
+            dictionary_factory, token_sampler
+        ).get_text_percentage_in_each_language(text, lang, greedy=True)
+
+    list_results = _convert_results_to_sorted_list(results)
+    # in case of ex-aequo use other token sampling to discriminate
+    if not greedy and list_results[0][1] == list_results[1][1]:
+        results = LanguageDetector(
+            dictionary_factory, backup_sampler
+        ).get_text_percentage_in_each_language(text, lang, greedy=True)
+
+    list_results = _convert_results_to_sorted_list(results)
+    return list_results
+
+
+class LanguageDetector:
+    __slots__ = ["dictionary_factory", "token_sampler"]
+
+    def __init__(
+        self,
+        dictionary_factory: DictionaryFactory = DictionaryFactory(),
+        token_sampler: TokenSampler = TokenSampler(),
+    ) -> None:
+        self.dictionary_factory = dictionary_factory
+        self.token_sampler = token_sampler
+
+    def get_text_percentage_in_each_language(
+        self,
+        text: str,
+        lang: Optional[Union[str, Tuple[str, ...]]] = None,
+        greedy: bool = False,
+    ) -> Dict[str, float]:
+        """Determine which proportion of the text is in the target language(s).
+        Perform a first run and further discriminate between the results if necessary.
+        """
+
+        tokens = self.token_sampler.sample_tokens(text)
+        total_tokens = len(tokens)
+        if total_tokens == 0:
+            return {"unk": 1}
+        # iterate
+        dictionaries = self.dictionary_factory.get_dictionaries(lang)
+        known_tokens_count = dict.fromkeys(dictionaries, 0)
```

Comment:
`defaultdict(int)` could be clearer here

Reply:
How is `defaultdict` any better than initializing a normal `dictionary` for all the languages starting at `0`?

Reply:
It's restricted to ints but it doesn't change much, we can leave it as it is now.